### ISSUE TITLE
Cannot get rich media working with 2.7.0 on IOS

### ISSUE THREAD

**Issue by rwillett** [OneSignal/OneSignal-Cordova-SDK#581]
**Description:**
We are struggling to get the Media Attachments working with 2.7.0. We had it working with previous versions of the OneSignal plugin on IOS 12 but since trying to upgrade to 2.6.0 and now 2.7.0 we cannot get it working with IOS on iPhone 13.2.2. This also coincides with a move to Xcode 11.2 to support IOS 13.2. 

We do get notifications come through BUT we do not get any attachment.

We have been through a wide number of steps to try and get this working ranging from downgrading to OneSignal 2.4.7 through to 2.7.0 on IOS.

**Environment**
We're now using OneSignal 2.7.0 with Xcode 11.2 (11B52).
Our plugins are:
```
com.rwillett.cordova.plugin.keychainutil 2.0.1 "KeyChain Plugin for Cordova iOS (No iCloud Sync)"
cordova-background-geolocation 2.10.1 "BackgroundGeolocation"
cordova-plugin-app-version 0.1.9 "AppVersion"
cordova-plugin-background-fetch 5.2.1 "CDVBackgroundFetch"
cordova-plugin-badge 0.8.7 "Badge"
cordova-plugin-battery-status 2.0.2 "Battery"
cordova-plugin-cocoalumberjack 0.0.4 "CocoaLumberjack"
cordova-plugin-device 1.1.2 "Device"
cordova-plugin-dialogs 2.0.1 "Notification"
cordova-plugin-email 1.2.6 "EmailComposer"
cordova-plugin-insomnia 4.2.0 "Insomnia (prevent screen sleep)"
cordova-plugin-network-information 1.1.0 "Network Information"
cordova-plugin-splashscreen 5.0.1 "Splashscreen"
cordova-plugin-sqlite-2 1.0.4 "SQLitePlugin"
cordova-plugin-statusbar 2.4.1 "StatusBar"
cordova-plugin-uniquedeviceid 1.3.2 "UniqueDeviceID"
cordova-plugin-vibration 3.0.1 "Vibration"
cordova-plugin-whitelist 1.3.3 "Whitelist"
cordova-plugin-x-toast 2.6.0 "Toast"
ionic-plugin-keyboard 2.2.1 "Keyboard"
london.jambuster.cordova.xcodedetector 0.1.2 "Xcode Detector"
onesignal-cordova-plugin 2.7.0 "OneSignal Push Notifications"
```
We have had to add 

<img width="319" alt="Screenshot 2019-11-11 at 09 45 52" src="https://user-images.githubusercontent.com/668923/68577478-185bca00-0468-11ea-99ba-75543e868d7b.png">

CoreGraphics and WebKit to get the 2.7.0 code to compile.

We have been through the steps in
https://documentation.onesignal.com/docs/cordova-sdk-setup
and also the steps in adding the Service Notification Extension 
https://documentation.onesignal.com/docs/ios-sdk-setup#section-1-add-notification-service-extension-recommend-
so many times we know it off by heart :)

We have edited the notificicationservice.m file so that it sends out a modified message and we can see the message on the Notification Drawer (or shade).
```
- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.receivedRequest = request;
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
    // [OneSignal didReceiveNotificationExtensionRequest:self.receivedRequest withMutableNotificationContent:self.bestAttemptContent];
    
    // DEBUGGING: Uncomment the 2 lines below and comment out the one above to ensure this extension is excuting
    //            Note, this extension only runs when mutable-content is set
    //            Setting an attachment or action buttons automatically adds this
    NSLog(@"Running NotificationServiceExtension");
    self.bestAttemptContent.body = [@"[Modified] " stringByAppendingString:self.bestAttemptContent.body];
    
    self.contentHandler(self.bestAttemptContent);
}
```
The body of the Notification alert has the word [Modified] at the start of it to show that is working (at least we got something right).

We have changed the log_level to 6 in our code but can't see any errors.

We have checked that we can see the PNG image that we are passing, this can be seen at 
https://preprod1.jambuster.london/Images/227166.png. This image is not likely to disappear in the next few days.

We have an exhaustive list of steps logged in EverNote as we keep trying new ideas. 

The only thing that comes to mind for us is IOS 13 and Xcode 11.2.

We have also been through 

`https://documentation.onesignal.com/docs/troubleshooting-cordova-variants

but that's not thrown any issues up.

We have tried pushing from the OneSignal dashboard and that works but no media.

Any help welcomed, the application throws no errors and looks like it should be working.

**Comment by rwillett** [comment]
We have also been through

https://documentation.onesignal.com/docs/notification-images-not-showing

and verified that all the suggestions have been done:

1.  Image URL is png or jpg
2. Image hosting is OK
3. OneSignal Image hosting N/A
4. Web Push images N/A
5. Checking that the Notification Service Extention is working. We can see that this is working as it has the [Modified] word inserted at the start of the body text,.
6. Checked we can directly open the PMNG image
7. The image is tiny and should only take a fraction of a second to download, not 30 secs.
8. Running IOS 13.2 <ahem>
9. URL is https.

**Comment by rwillett** [comment]
Checked that all IOS Deployment Targets are at the same level.

<img width="703" alt="Screenshot 2019-11-11 at 18 12 25" src="https://user-images.githubusercontent.com/668923/68610099-e3bf3100-04ae-11ea-8b4e-9152de942a46.png">


<img width="1000" alt="Screenshot 2019-11-11 at 18 13 41" src="https://user-images.githubusercontent.com/668923/68610140-02252c80-04af-11ea-9804-d2a7346c4c44.png">


<img width="1095" alt="Screenshot 2019-11-11 at 18 14 34" src="https://user-images.githubusercontent.com/668923/68610245-38fb4280-04af-11ea-99bd-e9ba4a3a3a71.png">


<img width="951" alt="Screenshot 2019-11-11 at 18 15 02" src="https://user-images.githubusercontent.com/668923/68610253-3dbff680-04af-11ea-9467-7900a6fecb99.png">


<img width="1004" alt="Screenshot 2019-11-11 at 18 15 51" src="https://user-images.githubusercontent.com/668923/68610283-529c8a00-04af-11ea-9132-b4d88424a48b.png">


<img width="1015" alt="Screenshot 2019-11-11 at 18 16 25" src="https://user-images.githubusercontent.com/668923/68610324-66e08700-04af-11ea-85d4-3d8f4d020ed3.png">

We've run out of ideas now and can't think what the problem is :(

**Comment by rgomezp** [comment]
Duplicate of https://github.com/OneSignal/OneSignal-iOS-SDK/issues/430

**Comment by rwillett** [comment]
I've read through #430 a few times but that relates to a bug in IOS12 and Xcode 10 from over a year ago. 

I also looked at the rdar for that bug report Oct 16th 2018 and that appears to be still open. 

Are you saying that a bug from Oct last year is still application in IOS 13 and Xcode 11?

Rob

**Comment by rgomezp** [comment]
@rwillett  you're looking at the wrong repository. make sure to click the link above

**Comment by rwillett** [comment]
I had read the correct link that you put in the reply. 

I put the wrong link in my response. I will edit it when back on laptop. 

My comment still stands. Your link refers to action buttons, rich media and xcode from over a year ago. Are you still saying there is an outstanding bug stopping rich media from working?

Thanks

Rob

**Comment by rgomezp** [comment]
@rwillett ,
You're correct. My apologies. I thought these were the same issue. 

What you're seeing isn't a known issue so that's interesting. Have you attempted reproducing on an older version of Xcode and/or iOS? Or have you only tested this with the iOS 13+ Xcode 11 combo?

**Comment by rwillett** [comment]
We have only tested this with ios 13+ and Xcode 11.

We have to use Xcode 11 to get access to IOS 13.2.2. We had a number of older phones with IOS 12 stolen a week or so ago, and we only have ios 13 phones left, we have a work iPhone 6 with IOS 12 but that's away until this weekend. My daughter has one but she may be reluctant to let us use it :)

We will try and use an old version of Xcode and ios 12, but I can't do it until the weekend. I will report back.

Do we need to look at adding in action buttons to see if that triggers anything?

Thanks

Rob

**Comment by rwillett** [comment]
This also appears to be similar to the bug I logged in January 

https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/474

Glad to see that Apple systems 'just work' :)

**Comment by rgomezp** [comment]
Yes, there is definitely a bug on Apple's side that must be fixed. We will be re-opening a ticket directly with Apple so we can try to get it resolved. 

Regarding this issue, let us know what you find.

**Comment by rwillett** [comment]
Ok, we have dug out (literally) an Iphone 6 to test:

1. iPhone is running 12.1.4 
2. OneSignal is 2.7.0
3. Xcode is 11.2 (11B52)

Our code base is the same as before.

We send a test message from the OneSignal console and it send the message but no image.

We have also tried sending it with action buttons. We managed to get this working in January (see https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/474) but that's not working now either.

We'll build a VMWare session with an older version of Xcode on and see if we can get it working on that.

Can you reproduce the issue there?

Thanks

Rob

**Comment by rwillett** [comment]
We can no confirm that the old version of our app thats on the app store still works correctly. So this means that the OneSignal push mechanism is working, so it puts the error onto 

1. IOS 13 
OR
2. Xcode 11.2
OR
3. OneSignal SDK 2.7.0

Our money is on IOS 13 :)

**Comment by rwillett** [comment]
We have downloaded and installed 

1. Xcode 10.3 (10G8) with 
2. iPhone 6 ios 12.1.4
3. OneSignal 2.7.0

and it doesn't work when sending from the OneSignal dashboard. We jave also set and unset Mutable content (I know we don't need to but we tried anyway).

So the next step is to roll OneSignal 2.7.0 back to 2.5.0. <sigh>.

Rob

**Comment by rwillett** [comment]
testing with Xcode 10.3 (10G8), IOS 12.1.4 and OneSignal.

OneSignal 2.5.0 doesn't work for us and neither does 2.4.7 which I believe is the last production one for us.

This makes me think it must be our code, but we haven't done much work in that area if anything. We've isolated everything else though so it must be our code.

Rob

**Comment by rwillett** [comment]
We've gone back to a release branch in git and created a new app.

We add that apps OneSignal ID to our test users segment in the OneSignal dashboard and send a rich media test message.  Lo and behold, it works. So we now have a version that we know works. 

This is with Xcode 10.3 (10G8)
IOS 12.1.4
OneSignal 2.4.7

The rest of our plugins are:

com.rwillett.cordova.plugin.keychainutil 2.0.1 "KeyChain Plugin for Cordova iOS (No iCloud Sync)"
cordova-background-geolocation 2.10.1 "BackgroundGeolocation"
cordova-plugin-app-version 0.1.9 "AppVersion"
cordova-plugin-background-fetch 5.2.1 "CDVBackgroundFetch"
cordova-plugin-badge 0.8.7 "Badge"
cordova-plugin-battery-status 2.0.2 "Battery"
cordova-plugin-cocoalumberjack 0.0.4 "CocoaLumberjack"
cordova-plugin-device 1.1.2 "Device"
cordova-plugin-dialogs 2.0.1 "Notification"
cordova-plugin-email 1.2.6 "EmailComposer"
cordova-plugin-insomnia 4.2.0 "Insomnia (prevent screen sleep)"
cordova-plugin-ionic-webview 1.1.16 "cordova-plugin-ionic-webview"
cordova-plugin-network-information 1.1.0 "Network Information"
cordova-plugin-splashscreen 5.0.1 "Splashscreen"
cordova-plugin-sqlite-2 1.0.4 "SQLitePlugin"
cordova-plugin-statusbar 2.4.1 "StatusBar"
cordova-plugin-uniquedeviceid 1.3.2 "UniqueDeviceID"
cordova-plugin-vibration 3.0.1 "Vibration"
cordova-plugin-whitelist 1.3.3 "Whitelist"
cordova-plugin-x-toast 2.6.0 "Toast"
ionic-plugin-keyboard 2.2.1 "Keyboard"
london.jambuster.cordova.xcodedetector 0.1.2 "Xcode Detector"
onesignal-cordova-plugin 2.4.7 "OneSignal Push Notifications"

Our intention now is to move the oneSignal plugin up one release at a time until we get to 2.7.0 (or it fails).

Rob

**Comment by rwillett** [comment]
This has not been forgotten, we are struggling to get anything to work on older released versions of code. See #586, until we can get any messages out, it's difficult to work on this :(

**Comment by rwillett** [comment]
As part of testing for this issue we have encountered version hell e.g. We can only get Xcode 11 for use with IOS 13.2.2, we decided to go straight to OneSignal 2.7.0, as we've had to move to cordova 5.0.1 we've had to make changes to code, ios 13.2.2 needs a later version of leaflet.js to handle multi-taps, etc, etc, etc. It's been a PITA TBH. 

Anyway we now dragged our app kicking and screaming into the 21st century and it now compiles and works on IOS 13.2.2 and IOS 12, but we still do **not** get any rich media passed down to either application on ios 12 or ios  13.

We've passed through lots of tests through the dashboard and can see the notification logged when the app is in the foreground. It even includes the image filename. The code preview below is a mixture of OneSignal LogLevel set to 6 and our own debugging. We have a standard set of handlers for IOS and Android so there functions in our code to capture the OneSignal Payload and configure it to our own generic format. It's pretty simple stuff and is readable (we hope),

```
2019-11-20 17:35:06.693135+0000 Jambuster[1551:556867] VERBOSE: onesignalUserNotificationCenter:willPresentNotification:withCompletionHandler: Fired!
2019-11-20 17:35:06.694582+0000 Jambuster[1551:556867] VERBOSE: notificationReceived:isActive called!
2019-11-20 17:35:06.694795+0000 Jambuster[1551:556867] VERBOSE: IAM Preview Not Detected, Handle Normal Notification
2019-11-20 17:35:06.695910+0000 Jambuster[1551:556867] VERBOSE: IAM Preview Not Detected, Handle Normal Notification
2019-11-20 17:35:06.696177+0000 Jambuster[1551:556867] VERBOSE: handleNotificationOpened:isActive called!
2019-11-20 17:35:06.697283+0000 Jambuster[1551:556867] VERBOSE: HTTP Request (OSRequestSubmitNotificationOpened) with URL: https://onesignal.com/api/v1/notifications/faed263e-fcdd-4cf7-9a36-fee9524d7476, with parameters: {
  "opened" : true,
  "app_id" : "434fbb2b-c79c-45d7-8628-cadbd154319c",
  "player_id" : "251f374d-36d8-47a4-ba45-93fab566db44"
}
2019-11-20 17:35:06.705501+0000 Jambuster[1551:556867] VERBOSE: callLegacyAppDeletegateSelector:withCompletionHandler: Fired!
2019-11-20 17:35:06.707026+0000 Jambuster[1551:556867] VERBOSE: oneSignalRemoteSilentNotification:UserInfo:fetchCompletionHandler:
2019-11-20 17:35:06.708355+0000 Jambuster[1551:556867] VERBOSE: notificationReceived:isActive called!
2019-11-20 17:35:06.732066+0000 Jambuster[1551:556867] HandleOneSignalIOSNotification: event = {
2019-11-20 17:35:06.732261+0000 Jambuster[1551:556867]   "payload": {
2019-11-20 17:35:06.732382+0000 Jambuster[1551:556867]     "rawPayload": {
2019-11-20 17:35:06.732518+0000 Jambuster[1551:556867]       "aps": {
2019-11-20 17:35:06.732693+0000 Jambuster[1551:556867]         "alert": {
2019-11-20 17:35:06.732806+0000 Jambuster[1551:556867]           "subtitle": "Test 1",
2019-11-20 17:35:06.732909+0000 Jambuster[1551:556867]           "title": "Test 1",
2019-11-20 17:35:06.733003+0000 Jambuster[1551:556867]           "body": "Test 1"
2019-11-20 17:35:06.733091+0000 Jambuster[1551:556867]         },
2019-11-20 17:35:06.733297+0000 Jambuster[1551:556867]         "mutable-content": 1,
2019-11-20 17:35:06.733413+0000 Jambuster[1551:556867]         "sound": "default"
2019-11-20 17:35:06.733489+0000 Jambuster[1551:556867]       },
2019-11-20 17:35:06.733582+0000 Jambuster[1551:556867]       "att": {
2019-11-20 17:35:06.733692+0000 Jambuster[1551:556867]         "id": "https://preprod1.jambuster.london/Images/227325.png"
2019-11-20 17:35:06.734844+0000 Jambuster[1551:556867]       },
2019-11-20 17:35:06.735004+0000 Jambuster[1551:556867]       "custom": {
2019-11-20 17:35:06.735160+0000 Jambuster[1551:556867]         "i": "faed263e-fcdd-4cf7-9a36-fee9524d7476"
2019-11-20 17:35:06.735273+0000 Jambuster[1551:556867]       }
2019-11-20 17:35:06.735376+0000 Jambuster[1551:556867]     },
2019-11-20 17:35:06.735475+0000 Jambuster[1551:556867]     "subtitle": "Test 1",
2019-11-20 17:35:06.735872+0000 Jambuster[1551:556867]     "title": "Test 1",
2019-11-20 17:35:06.735991+0000 Jambuster[1551:556867]     "sound": "default",
2019-11-20 17:35:06.736087+0000 Jambuster[1551:556867]     "body": "Test 1",
2019-11-20 17:35:06.736204+0000 Jambuster[1551:556867]     "notificationID": "faed263e-fcdd-4cf7-9a36-fee9524d7476",
2019-11-20 17:35:06.736316+0000 Jambuster[1551:556867]     "actionButtons": []
2019-11-20 17:35:06.736512+0000 Jambuster[1551:556867]   },
2019-11-20 17:35:06.737250+0000 Jambuster[1551:556867]   "shown": false,
2019-11-20 17:35:06.737412+0000 Jambuster[1551:556867]   "isAppInFocus": true
2019-11-20 17:35:06.737525+0000 Jambuster[1551:556867] }
2019-11-20 17:35:06.737598+0000 Jambuster[1551:556867] event.silentNotification has NOT been set
2019-11-20 17:35:06.737668+0000 Jambuster[1551:556867] HandleOneSignalIOSNotification: iOSConvertedOneSignalNotification = {
2019-11-20 17:35:06.737765+0000 Jambuster[1551:556867]   "contentAvailable": false,
2019-11-20 17:35:06.738804+0000 Jambuster[1551:556867]   "message": "Test 1",
2019-11-20 17:35:06.738939+0000 Jambuster[1551:556867]   "additionalData": {
2019-11-20 17:35:06.739050+0000 Jambuster[1551:556867]     "Disruption_Id": -1,
2019-11-20 17:35:06.739147+0000 Jambuster[1551:556867]     "EpochTime": -1,
2019-11-20 17:35:06.739308+0000 Jambuster[1551:556867]     "Foreground": true,
2019-11-20 17:35:06.739405+0000 Jambuster[1551:556867]     "onStart": false
2019-11-20 17:35:06.749883+0000 Jambuster[1551:556867] THREAD WARNING: ['Console'] took '10.415039' ms. Plugin should use a background thread.
2019-11-20 17:35:06.750246+0000 Jambuster[1551:556867]   },
2019-11-20 17:35:06.750436+0000 Jambuster[1551:556867]   "userClicked": false
2019-11-20 17:35:06.750531+0000 Jambuster[1551:556867] }
2019-11-20 17:35:06.929792+0000 Jambuster[1551:557315] VERBOSE: network response (OSRequestSubmitNotificationOpened): {
    success = 1;
}
```

We are running out of ideas now, its very difficult for us to get Xcode at a lower level as we need Xcode 11. As it doesn't work on IOS 12 and IOS 13, that rules out IOS 13 as the problem. This application does work with older versions and we haven't updated the OneSignal notifications code.

I'd hate for us to put out an upgraded app for IOS 13 which has a downgraded notification section :(

Rob
Any other ideas welcomed

**Comment by rwillett** [comment]
We have just upgraded to Xcode Version 11.2.1 (11B500) and the issue is sadly still there.

We have also distributed this build via Testflight to check if that has anything to do with it, and that didn't change anything.

Has anybody else had this problem and is it reproducible?

Thanks

Rob

**Comment by rgomezp** [comment]
So to summarize, you’ve tested on an old version of Xcode, iOS, and OneSignal and it still doesn’t work. This means it’s likely a problem with your code. You then updated to the latest iOS and OneSignal SDK, resolved version issues, but rich media is still not working. Correct?

It doesn't seem other folks are encountering this. Perhaps the issue is in your service extension code? Could you post it here?

**Comment by rwillett** [comment]
We can get it working on an older released version of OneSignal, 2.4.7. However due to the fact we have ios 13.2, we had to go to later versions of OneSignal and also we have had to go to Xcode 11.2 and then Xcode 11.2.1. We are in version hell. It does work with 2.4.7 on older versions of IOS though and we haven't changed our code.

Here's the 

```
#import <OneSignal/OneSignal.h>

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNNotificationRequest *receivedRequest;
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.receivedRequest = request;
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
    // [OneSignal didReceiveNotificationExtensionRequest:self.receivedRequest withMutableNotificationContent:self.bestAttemptContent];
    
    // DEBUGGING: Uncomment the 2 lines below and comment out the one above to ensure this extension is excuting
    //            Note, this extension only runs when mutable-content is set
    //            Setting an attachment or action buttons automatically adds this
    NSLog(@"Running NotificationServiceExtension");
    self.bestAttemptContent.body = [@"[X-Modified] " stringByAppendingString:self.bestAttemptContent.body];
    
    self.contentHandler(self.bestAttemptContent);
}

- (void)serviceExtensionTimeWillExpire {
    // Called just before the extension will be terminated by the system.
    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
    
    [OneSignal serviceExtensionTimeWillExpireRequest:self.receivedRequest withMutableNotificationContent:self.bestAttemptContent];
    
    self.contentHandler(self.bestAttemptContent);
}

@end
```

This is (almost) straight out of your website. We changed the 'modified' in the debug to x-modified.

We have also been through the issues both open and closed to see if anybody else has had the same issue. The main one seems to be making sure that all the versions are right, thats why we posted the pics

We have been through the OneSignal documents carefully and logged each line by line as we do them. We don't get any errors when we send a message, we get the notification but no image.

Thanks

Rob

**Comment by rwillett** [comment]
We have checked the documentation again and for xcode 11 , IOS 13 and for OneSIgnal 2.7.0 we need to manually add in extra libraries that are not in your instructions

<img width="280" alt="Screenshot 2019-11-27 at 09 28 55" src="https://user-images.githubusercontent.com/668923/69711180-8525c400-10f8-11ea-9e24-779e0b2af7c0.png">

CoreGraphics
WebKit

Could this have anything to do with the issue?

Thanks

Rob

**Comment by rwillett** [comment]
We have downloaded and built the OneSignal Ionic 2 sample app and the example works. It sends an image from the phone which is displayed. Also if we send a test message from the OneSignal dashboard then it works.

Since we followed the exact same docs as before, it does seem to indicate the problem is in our code or our build.

Rob

**Comment by rgomezp** [comment]
I will update the instructions. Thanks for bringing to my attention

**Comment by rwillett** [comment]
We have now got the application working but have no real idea how and why.

We created a new Ionic 1 project and reversed our code into that project, installed all the plugins and platforms, added the OneSignal extension (in exactly the same way as we have done > 20 times previously) and it works. We didn't change a line of our code. We also have the same plugins and platforms at the same revisions as before. 

We did reinstall cocapods and node/npm just in case. 

We can see rich media now in notifications but we can't see what is different with this build, but clearly something is.

I'm closing this as I can't see us finding anything else out about the problem.
